### PR TITLE
com.lunarclient.LunarClient

### PR DESCRIPTION
**Pull request by Doomsdayrs**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. image
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: com.lunarclient.Launcher.appdata.xml
Comment (hfiguiere) [inline com.lunarclient.Launcher.appdata.xml:66]: screenshots need to be accessible via https.

File: com.lunarclient.Launcher.yml
Comment (hfiguiere) [inline com.lunarclient.Launcher.yml:54]:
```suggestion
      - install  -Dm644 -t ${FLATPAK_DEST}/share/icons/hicolor/256x256/apps install/lunarclient.png
```

File: com.lunarclient.Launcher.appdata.xml
Comment (hfiguiere) [inline com.lunarclient.Launcher.appdata.xml:33]: appstream validate will not like bare URL. URL should be in the `<url></url>` tag instead. (you can skip having a `<desription>`)

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by hfiguiere** [comment]
bot, build com.lunarclient.Launcher

**Comment by nedrichards** [comment]
bot, build com.lunarclient.Launcher

File: com.lunarclient.Launcher.yml
Comment (hfiguiere) [inline com.lunarclient.Launcher.yml:53]:
```suggestion
      - install  -Dm644 -t ${FLATPAK_DEST}/share/icons/hicolor/256x256/apps icons/lunarclient.png
```

Gosh, I messed up the original suggestion.

**Comment by surfaceflinger** [comment]
Upstream might be hostile towards this idea. Lunar Client is actual spyware/malware and proof is [here](https://www.lunarclient.com/privacy/).

What we're doing here is repacking their AppImage and preventing them from doing the spyware thing.

> You may not access, copy, modify or distribute the Services, unless expressly authorized by Company or permitted by         applicable law.
> You may not reverse engineer or attempt to extract or otherwise use source code or other data from the Services, unless expressly authorized by Company or permitted by law.
> Company or its licensors own and reserve all other rights, including all right, title and interest in the Services and all materials or content therein, and all associated Intellectual Property Rights.

Even simply asking upstream about permissions for flathub redistribution might lead to them adding some kind of sandbox prevention.

**Comment by Doomsdayrs** [comment]
>  You may not reverse engineer or attempt to extract or otherwise use source code or other data from the Services, unless expressly authorized by Company or permitted by law.

We are not extracting source code tho?

**Comment by hfiguiere** [comment]
> Upstream might be hostile towards this idea. Lunar Client is actual spyware/malware and proof is [here](https://www.lunarclient.com/privacy/).

Whell `   - --share=network`, so much for sandboxing a "spyware".

**Comment by Doomsdayrs** [comment]
bot, build com.lunarclient.LunarClient

**Comment by hfiguiere** [comment]
AppData fail validation:

```
builddir/files/share/appdata/com.lunarclient.LunarClient.appdata.xml: FAILED:
• tag-invalid           : <project_license> is not valid [Proprietary]SPDX ID 'Proprietary' unknown
• style-invalid         : <release> description should be prose and not contain hyperlinks [<p>Hey everyone, our primary focus these past few weeks has been to fix bugs, implement new mods, and add quality of life features!</p><p>We finally have the much-requested Quickplay Mod. This mod allows you to switch games on the Hypixel Network seamlessly. We also added a new mod called “Snaplook” that will enable you to change/toggle your perspective by pressing a set key (no, this is not freelook). Lastly, we added the 3D Skins mod that allows you to add more details to your extra skin layers. Along with adding these new mods, we completely revamped the Crosshair Mod, fixed some much-needed bugs, and improved various aspects of our UI. Oh – we also have Discord linking! If you aren’t yet in our Discord, make sure to join!</p><p>Also, we are aware of launching issues on Intel macOS on versions 1.7, 1.8, and 1.12. We are actively working on a solution and should have an update for this soon, sorry for the inconvenience.</p><p>If you have any suggestions to improve Lunar Client, make sure to interact with us on our social media accounts (https://www.lunarclient.com/socials).</p><p>We are still actively searching for new hires. Specifically, we are looking for new client developers to implement and create new Hypixel Skyblock mods. Visit our jobs page (https://www.lunarclient.com/jobs) for all relevant information. Thank you all for the constant support; Lunar Client is made for the players, by the players.</p>]
```

File: com.lunarclient.LunarClient.appdata.xml
Comment (dada513) [inline com.lunarclient.LunarClient.appdata.xml:76]: should be `<project_license>LicenseRef-proprietary</project_license>`

Comment (hfiguiere) [reply com.lunarclient.LunarClient.appdata.xml:76]: yes, see https://github.com/flathub/flathub/pull/2784#issuecomment-1018858368

**Comment by Doomsdayrs** [comment]
bot, build com.lunarclient.LunarClient

**Comment by Doomsdayrs** [comment]
:D

**Comment by rowbawts** [comment]
I tried that latest successful build and this is the error I get when I try and launch the game.

Screenshot from 2022-01-22 18-36-47

**Comment by dada513** [comment]
> Edit: I tried it again and it worked now that was weird.
> 
> ~I tried that latest successful build and this is the error I get when I try and launch the game.~
> 
> Screenshot from 2022-01-22 18-36-47

I got it too, If you restart the it it works might be an error on their side

**Comment by Doomsdayrs** [comment]
I cleaned up the appdata and other hinks. This should be the last one before ready
bot, build com.lunarclient.LunarClient

**Comment by Doomsdayrs** [comment]
bot, build com.lunarclient.LunarClient

**Comment by dada513** [comment]
Is bot broken?

**Comment by hfiguiere** [comment]
it's in queue.

**Comment by Doomsdayrs** [comment]
For over 4 hours?

**Comment by hfiguiere** [comment]
you are not alone

**Comment by nedrichards** [comment]
You can always check build status on https://flathub.org/builds/ (when the app is accepted you'll be able to do more stuff by logging in there too)

**Comment by hfiguiere** [comment]
the queue was clogged. It's clearing up now.

**Comment by Doomsdayrs** [comment]
It failed, because it was too long???

**Comment by hfiguiere** [comment]
yup

**Comment by hfiguiere** [comment]
You can install `org.freedesktop.appstream-glib` to validate the file on the command line

**Comment by Doomsdayrs** [comment]
bot, build com.lunarclient.LunarClient

File: com.lunarclient.LunarClient.yml
Comment (barthalion) [inline com.lunarclient.LunarClient.yml:35]: I still suggest you ask upstream for redistribution rights.

File: com.lunarclient.LunarClient.yml
Comment (barthalion) [inline com.lunarclient.LunarClient.yml:36]: This should be configured to detect new versions with https://github.com/flathub/flatpak-external-data-checker/.

File: com.lunarclient.LunarClient.yml
Comment (barthalion) [inline com.lunarclient.LunarClient.yml:17]: Does it actually support Wayland?

File: com.lunarclient.LunarClient.yml
Comment (barthalion) [inline com.lunarclient.LunarClient.yml:4]:
```suggestion
```

File: com.lunarclient.LunarClient.yml
Comment (dada513) [reply com.lunarclient.LunarClient.yml:17]: It does not at least without a custom glfw which this launcher does not allow to be used

File: com.lunarclient.LunarClient.yml
Comment (dada513) [reply com.lunarclient.LunarClient.yml:36]: Need YAML checker, impossible until https://github.com/flathub/flatpak-external-data-checker/issues/268

File: com.lunarclient.LunarClient.yml
Comment (surfaceflinger) [reply com.lunarclient.LunarClient.yml:35]: @phase which is one Lunar devs already left a "hooray" reaction on the main PR comment so I guess they're fine with it. It would be definitely good if they could leave a comment instead here though.

File: com.lunarclient.LunarClient.yml
Comment (surfaceflinger) [reply com.lunarclient.LunarClient.yml:17]: Launcher should support Wayland because it's using Chromium. Minecraft itself doesn't support it, but it's possible to swap  glfw x11 for Wayland one. It should be possible for upstream to include patched glfw in Launcher and detect if Wayland is available. https://github.com/Admicos/minecraft-wayland

File: com.lunarclient.LunarClient.yml
Comment (Doomsdayrs) [reply com.lunarclient.LunarClient.yml:35]: > I still suggest you ask upstream for redistribution rights.

Why?

File: com.lunarclient.LunarClient.yml
Comment (dada513) [inline com.lunarclient.LunarClient.yml:39]:
```suggestion
        filename: lunar-client.appimage
```

File: apply_extra.sh
Comment (dada513) [inline apply_extra.sh:3]:
```suggestion
APP_IMAGE="/app/extra/lunar-client.appimage"
```

**Review by dada513 — COMMENTED** [review]
I suggest removing the version number to make it easier to update once external-data-checker supports YAML.
That way (after its supported) you could just merge the flathubbot PR and have it updated instead of having an old version number in the manifest (confusing for anyone that will read the manifest)

File: com.lunarclient.LunarClient.yml
Comment (Doomsdayrs) [reply com.lunarclient.LunarClient.yml:35]: This can be done using rotating-url which can work with app images.

https://github.com/flathub/flatpak-external-data-checker#url-checker

```
If the filename ends with .AppImage, the version number is extracted from the AppImage. (It is run in a bwrap sandbox.)
```

**Comment by dada513** [comment]
I don't think the x-checker-data you added will work, rotating URL is for when the vendor has a redirect to the latest version (I think). Like I mentioned before Lunar uses a yml file for this, which currently isn't supported by f-e-d-c

File: com.lunarclient.LunarClient.yml
Comment (dada513) [reply com.lunarclient.LunarClient.yml:35]: > This can be done using rotating-url which can work with app images.
> 
> https://github.com/flathub/flatpak-external-data-checker#url-checker
> 
> ```
> If the filename ends with .AppImage, the version number is extracted from the AppImage. (It is run in a bwrap sandbox.)
> ```

Ah, may be it. Ignore my last comment as I didn’t see notification for this

File: com.lunarclient.LunarClient.yml
Comment (dada513) [reply com.lunarclient.LunarClient.yml:34]: > > I still suggest you ask upstream for redistribution rights.
> 
> Why?

I think faster downloads, not having broken links and delta updates are the main reasons for the flathub CDN being preffered

File: com.lunarclient.LunarClient.yml
Comment (dada513) [reply com.lunarclient.LunarClient.yml:35]: yeah it doesn't work
```
ERROR   src.checker: Error reading source: [ordereddict([('type', 'rotating-url'), ('url', 'https://launcherupdates.lunarclientcdn.com/Lunar%20Client-2.9.3.AppImage')])] is not of type 'object'

Failed validating 'type' in schema['properties']['x-checker-data']:
    {'properties': {'arches': {'items': {'type': 'string'},
                               'type': 'array'},
                    'is-main-source': {'type': 'boolean'},
                    'type': {'type': 'string'}},
     'required': ['type'],
     'type': 'object'}

On instance['x-checker-data']:
    [ordereddict([('type', 'rotating-url'), ('url', 'https://launcherupdates.lunarclientcdn.com/Lunar%20Client-2.9.3.AppImage')])]
INFO    src.checker: Checking 2 external data items
INFO    src.checker: Started check [1/2] git unappimage (from com.lunarclient.LunarClient.yml)
INFO    src.checkers.gitchecker: Skipping source unappimage, commit is specified, but neither tag nor branch
INFO    src.checker: Finished check [1/2] git unappimage (from com.lunarclient.LunarClient.yml)
INFO    src.checker: Skipped check [2/2] archive xrandr-1.5.1.tar.gz (from com.lunarclient.LunarClient.yml)
WARNING src.main: Check finished with 1 error(s)
```
I believe the appimage version is only used to construct a new version-specific url using the vendor's latest release url. Not to get the latest version from an old appimage somehow

Comment (dada513) [reply com.lunarclient.LunarClient.yml:35]: a new f-e-d-c PR was made with the required changes: [#270](https://github.com/flathub/flatpak-external-data-checker/pull/270)

**Comment by nedrichards** [comment]
bot, build com.lunarclient.LunarClient

File: com.lunarclient.LunarClient.yml
Comment (gasinvein) [inline com.lunarclient.LunarClient.yml:40]: `x-checker-data` should be an object, not an array.

File: com.lunarclient.LunarClient.yml
Comment (Doomsdayrs) [reply com.lunarclient.LunarClient.yml:34]: I do not think it is a problem currently.

The flatpak works & downloads fine.

**Comment by Doomsdayrs** [comment]
bot, build com.lunarclient.LunarClient

File: com.lunarclient.LunarClient.yml
Comment (barthalion) [reply com.lunarclient.LunarClient.yml:34]: "Fine" is really a poor evaluation here. By stubbornly sticking to extra-data, it's losing delta updates and a proper CDN setup for faster downloads. It's great you're happy with it but it could be done better if you could simply clarify that with upstream.

Comment (Doomsdayrs) [reply com.lunarclient.LunarClient.yml:34]: I sent your request to the developers.

Comment (Doomsdayrs) [reply com.lunarclient.LunarClient.yml:34]: No reply on that, I do not want this holding up users from using this application that is working

**Comment by phase** [comment]
Hello, Lunar Client developer here.

image
No one has redistribution rights for Lunar Client.

---

> Upstream might be hostile towards this idea. Lunar Client is actual spyware/malware and proof is [here](https://www.lunarclient.com/privacy/).

Lunar Client is not spyware. It's a mod pack for Minecraft, much like Minecraft Forge or Fabric with a small set of mods bundled in. Our terms of service are not "proof" of any wrongdoing.

---

Overall I don't think this is the best idea. We are constantly pushing out launcher updates, and maintaining backwards compatibility with old launcher versions is a pain for us. The launcher auto-updates so we don't have people using these older versions. Having many people download this flatpack and letting it rot would mean they either a) never update again and quit Lunar Client, or b) have to download our launcher manually anyway.

**Comment by Doomsdayrs** [comment]
@phase 
Thank you for your reply!

---

I hope you can dismiss the users who claim LunarClient to be spyware, as they are not affiliated with me.

---

I do not intend to let this launcher rot on flathub, as one of the requirements is to constantly maintain the flatpak application. Which would be something intend to uphold, as my little brother uses lunar client heavily, and I have an interest to keep it running the best as possible. 

If that does not seem like something you would wish, you can also take control of the flatpak and provide a spectacular experience for the Linux user base.

File: com.lunarclient.LunarClient.yml
Comment (barthalion) [inline com.lunarclient.LunarClient.yml:40]: Have you tested it that it actually works? Looking at the example in the README and usage on Flathub at https://github.com/search?q=org%3Aflathub+rotating-url&type=code, I sincerely doubt it.

**Comment by rowbawts** [comment]
> Hello, Lunar Client developer here.
> 
> image
> No one has redistribution rights for Lunar Client.
> 
> > Upstream might be hostile towards this idea. Lunar Client is actual spyware/malware and proof is [here](https://www.lunarclient.com/privacy/).
> 
> Lunar Client is not spyware. It's a mod pack for Minecraft, much like Minecraft Forge or Fabric with a small set of mods bundled in. Our terms of service are not "proof" of any wrongdoing.
> 
> Overall I don't think this is the best idea. We are constantly pushing out launcher updates, and maintaining backwards compatibility with old launcher versions is a pain for us. The launcher auto-updates so we don't have people using these older versions. Having many people download this flatpack and letting it rot would mean they either a) never update again and quit Lunar Client, or b) have to download our launcher manually anyway.

Yes I'd also like to express my use case for this to be published as my distro (Silverblue) only *really* supports Flatpaks as its package manager so having this would allow me to use your launcher as I'd like to.

File: com.lunarclient.LunarClient.yml
Comment (dada513) [reply com.lunarclient.LunarClient.yml:40]: Use the electron checker. This won't work

Comment (Doomsdayrs) [reply com.lunarclient.LunarClient.yml:40]: Are you sure about the electron checker?

**Comment by Doomsdayrs** [comment]
With the release of the SteamDeck, being considerate to the SteamDeck users by supporting the flatpak release would be very awesome :D.

File: com.lunarclient.LunarClient.yml
Comment (dada513) [reply com.lunarclient.LunarClient.yml:40]: Yes. It's made specifically for apps like this

Rotating url only checks for a link redirecting to the latest version, this isn't the case here

**Comment by dada513** [comment]
> letting it rot would mean they either a) never update again and quit Lunar Client, or b) have to download our launcher manually anyway.

Not only did the publisher say they won't let it rot, but with the version checker setup it should be nearly automatic to update it as long as there are no breaking changes for flatpak on upstream's part

File: com.lunarclient.LunarClient.yml
Comment (barthalion) [reply com.lunarclient.LunarClient.yml:40]: Any luck fixing this?

Comment (dada513) [reply com.lunarclient.LunarClient.yml:40]: Yes, if author switches to electron checker

Comment (barthalion) [reply com.lunarclient.LunarClient.yml:40]: Which is the point of my question.

Comment (Doomsdayrs) [reply com.lunarclient.LunarClient.yml:40]: Running
```yml
url: https://launcherupdates.lunarclientcdn.com/Lunar%20Client-2.9.4.AppImage
        size: 87798866 
        sha256: de1d8516920846d6ba844b1afc7d31a3cf8352f85d43206ff5da9d77fbe5c19e
        filename: lunar-client.appimage
        x-checker-data:
            type: electron-updater
```

Does in fact provide use with an updated URL
```
State.LATEST|BROKEN: lunar-client.appimage
 Has a new version:
  URL:       https://launcherupdates.lunarclientcdn.com/Lunar%20Client-2.10.1.AppImage
  MD5:       d4eef58ee5758ae081af0c3f35ecccf7
  SHA1:      2b350cb2d0bfd2e8b2e0a907c956935331dae6cd
  SHA256:    de1d8516920846d6ba844b1afc7d31a3cf8352f85d43206ff5da9d77fbe5c19e
  SHA512:    77b39e542c00ce9ebad9d2921a6c8f35314a025333c6607703f7308756d81df82bda89b53f754e44ed8329e6ea3c9ee5d79e28a7db49ac869efdecbd36b846dc
  Size:      87774312
  Version:   2.10.1
  Timestamp: 2022-04-07 19:48:25.342000+00:00
  ```
  
  But change it to 2.10.1, and it will *still* pump out that there is an update. @rbrick Do you think you can educate us on how to let flatpak check for an update, so it can automatically update the launcher? A url or anything that we can use?

Comment (dada513) [reply com.lunarclient.LunarClient.yml:40]: I am pretty sure it does that because you didn't copy the sha256 in updated one. I had that issue. What if you let it update it not only check?

Comment (Doomsdayrs) [reply com.lunarclient.LunarClient.yml:40]: I just double tested it, Same result. Try it yourself too.

It keeps spitting out that there is an update

Comment (DaPigGuy) [reply com.lunarclient.LunarClient.yml:40]: image

You might be forgetting to update the URL? I have not been able to reproduce any issues when changing both the URL and sha256 hahs.

Comment (Doomsdayrs) [reply com.lunarclient.LunarClient.yml]: Following @DaPigGuy pointer, I have realized the error of my ways.

The FEDC does work

**Comment by Doomsdayrs** [comment]
bot, build com.lunarclient.Launcher

**Comment by Doomsdayrs** [comment]
bot, build com.lunarclient.LunarClient

**Comment by Doomsdayrs** [comment]
Updated with direct proof of permission

**Comment by Doomsdayrs** [comment]
> > Updated with direct proof of permission
> 
> Where is the direct proof?

In the original post

**Pull request closed without merging**
